src/realm/object_converter.cpp

```diff
@@ -505,73 +514,83 @@ void InterRealmValueConverter::handle_dictionary_in_mixed(Dictionary& src_dictio
 }
 
 bool InterRealmValueConverter::check_matching_list(const Lst<Mixed>& src_list, Lst<Mixed>& dst_list, size_t ndx_src,
-                                                   size_t ndx_dst, CollectionType type) const
+                                                   size_t ndx_dst) const
 {
-
-    if (type == CollectionType::List) {
-        auto nested_src_list = src_list.get_list(ndx_src);
-        auto nested_dst_list = dst_list.get_list(ndx_dst);
-        auto size_src = nested_src_list->size();
-        auto size_dst = nested_dst_list->size();
-        if (size_src != size_dst)
+    REALM_ASSERT(ndx_src < src_list.size() && ndx_dst < dst_list.size());
+    auto src_any = src_list.get_any(ndx_src);
+    auto dst_any = dst_list.get_any(ndx_dst);
+    if (src_any != dst_any)
+        return false;
+    if (!is_collection(src_any))
+        return true;
+    // Check collections are equal
+    if (src_any.is_type(type_List)) {
+        auto src_element = src_list.get_list(ndx_src);
+        auto dst_element = dst_list.get_list(ndx_dst);
+        auto src_element_size = src_element->size();
+        auto dst_element_size = dst_element->size();
+        if (src_element_size != dst_element_size)
             return false;
-        for (size_t i = 0; i < size_src; ++i) {
-            auto src_mixed = nested_src_list->get_any(i);
-            auto dst_mixed = nested_dst_list->get_any(i);
-            if (src_mixed != dst_mixed)
+        for (size_t i = 0; i < src_element_size; ++i) {
+            if (!(check_matching_list(*src_element, *dst_element, i, i)))
                 return false;
         }
     }
-    else if (type == CollectionType::Dictionary) {
-        auto nested_src_dictionary = src_list.get_dictionary(ndx_src);
-        auto nested_dst_dictionary = dst_list.get_dictionary(ndx_dst);
-        auto size_src = nested_src_dictionary->size();
-        auto size_dst = nested_dst_dictionary->size();
-        if (size_src != size_dst)
+    else if (src_any.is_type(type_Dictionary)) {
+        auto src_element = src_list.get_dictionary(ndx_src);
+        auto dst_element = dst_list.get_dictionary(ndx_dst);
+        auto src_element_size = src_element->size();
+        auto dst_element_size = dst_element->size();
+        if (src_element_size != dst_element_size)
             return false;
-        for (size_t i = 0; i < size_src; ++i) {
-            auto [src_key, src_mixed] = nested_src_dictionary->get_pair(i);
-            auto [dst_key, dst_mixed] = nested_dst_dictionary->get_pair(i);
+        for (size_t i = 0; i < src_element_size; ++i) {
+            auto src_key = src_element->get_key(i);
+            auto dst_key = dst_element->get_key(i);
             if (src_key != dst_key)
                 return false;
-            if (src_mixed != dst_mixed)
+            if (!check_matching_dictionary(*src_element, *dst_element, src_key.get_string()))
```

Comment:
Dictionaries can have a key type of Int as well as string. I think there are several places that you should be passing around a mixed for the key, instead of a string. (this appears to be a preexisting problem). We should also test what happens when recovering dictionary instructions that apply to a dictionary that has had a key type change.

Reply:
indeed. I remember we had some discussions about allowing int keys (and maybe other types). I could make the change if we think it's important, otherwise we can do it at a later point when we add full support for it.